## Diploma UCU - Proyecto de Análisis de Datos con Python - Nicolás Gallo

En el presente proyecto, el objetivo es poder analizar datos de Egresos Hospitalarios en Uruguay entre los años 2021 y 2024.
El dataset fue obtenido del Catálogo de Datos Abierto y se seleccionaron estos años para el análisis por ser los datos disponibles más actuales a la fecha.

En el dataset, se encuentran datos de los egresos hospitalarios en Uruguay, significando esto casos en los que la persona abandonó el centro hospitalario pudiendo ser consecuencia de su fallecimiento o de su alta domiciliaria, sin er ello explícitamente identificado. Dentro de los datos, se encuentra no solamente el año, sino también el género, región, grupo etareo, sistema de salud, lo que permite cruzarlos para lograr obtener información relevante sobre tendencias.

Comenzando con la fase 1 del trabajo, importo las librerías a utilizar y traigo el dataset intervenido para proceder a su EDA.
Presento situación primaria del dataset y contabilizo los valores nulos que serán limpiados.

In [14]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

base_dir = os.path.abspath(os.path.join(os.getcwd(), '..'))
ruta = os.path.join(base_dir, 'data', 'raw', 'p4ds_grupo_c_egresos_hospitalarios.csv')

print("Buscando en:", ruta)

df = pd.read_csv(ruta, sep=';', dtype_backend='numpy_nullable')

display(df.head())
print(f"Filas: {df.shape[0]:,} | Columnas: {df.shape[1]}")
df.info()
print("\n- VALORES NULOS -")
print(df.isnull().sum())

Buscando en: C:\Users\usuario\PycharmProjects\nicolas_gallo_diploma_proyecto_python\data\raw\p4ds_grupo_c_egresos_hospitalarios.csv


,ID,AÑO,SECTOR,GENERO,GRUPO ETAREO,REGION,DIAGNOSTICO,CAUSA EXTERNA
0,1,2024,PRIVADO,F,65+,INTERIOR,Hemorragia intracraneal,NC
1,2,2024,PRIVADO,M,65+,INTERIOR,Tumor maligno de la vejiga urinaria,NC
2,3,2024,PRIVADO,M,65+,INTERIOR,Otras enfermedades del sistema urinario,NC
3,4,2024,PRIVADO,F,65+,INTERIOR,Otras enfermedades del sistema urinario,NC
4,5,2024,PRIVADO,M,65+,INTERIOR,Otras enfermedades del sistema urinario,NC


Filas: 1,132,011 | Columnas: 8
<class 'pandas.DataFrame'>
RangeIndex: 1132011 entries, 0 to 1132010
Data columns (total 8 columns):
 #   Column         Non-Null Count    Dtype 
---  ------         --------------    ----- 
 0   ID             1132011 non-null  Int64 
 1   AÑO            1132011 non-null  Int64 
 2   SECTOR         1132011 non-null  string
 3   GENERO         1132011 non-null  string
 4   GRUPO ETAREO   1132011 non-null  string
 5   REGION         1089284 non-null  string
 6   DIAGNOSTICO    1132011 non-null  string
 7   CAUSA EXTERNA  1132011 non-null  string
dtypes: Int64(2), string(6)
memory usage: 146.5 MB

- VALORES NULOS -
ID                   0
AÑO                  0
SECTOR               0
GENERO               0
GRUPO ETAREO         0
REGION           42727
DIAGNOSTICO          0
CAUSA EXTERNA        0
dtype: int64


Presento los valores de cada columna que son únicos, para identificar también posibilidades futuras de visualización.

In [15]:
print("- VALORES ÚNICOS POR COLUMNA -")
for col in df.columns:
    print(f"\n{col} ({df[col].nunique()} únicos):")
    print(df[col].unique()[:10])  # primeros 10 valores únicos

- VALORES ÚNICOS POR COLUMNA -

ID (1048158 únicos):
<IntegerArray>
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Length: 10, dtype: Int64

AÑO (4 únicos):
<IntegerArray>
[2024, 2023, 2022, 2021]
Length: 4, dtype: Int64

SECTOR (2 únicos):
<ArrowStringArray>
['PRIVADO', 'PUBLICO']
Length: 2, dtype: string

GENERO (2 únicos):
<ArrowStringArray>
['F', 'M']
Length: 2, dtype: string

GRUPO ETAREO (10 únicos):
<ArrowStringArray>
[      '65+',   '45 a 64',   '25 a 44',   '20 a 24',   '15 a 19',   '10 a 14',
     '5 a 9',     '1 a 4', 'Menor a 1', 'Sin datos']
Length: 10, dtype: string

REGION (4 únicos):
<ArrowStringArray>
['INTERIOR', 'MONTEVIDEO', <NA>, 'EXTERIOR', 'Sin datos']
Length: 5, dtype: string

DIAGNOSTICO (258 únicos):
<ArrowStringArray>
[                     'Hemorragia intracraneal',
          'Tumor maligno de la vejiga urinaria',
      'Otras enfermedades del sistema urinario',
                           'Fractura del fémur',
                     'Traumatismo intracraneal',
               

Identifico filas que se encuentran dobles en el dataset para poder limpiarlas y se muestran algunas de estas filas.

In [16]:
print("- DUPLICADOS -")
print(f"Filas duplicadas: {df.duplicated().sum():,}")
print(f"IDs duplicados: {df['ID'].duplicated().sum():,}")

- DUPLICADOS -
Filas duplicadas: 83,853
IDs duplicados: 83,853


In [17]:
duplicados = df[df.duplicated(keep=False)]
print(f"Total de filas involucradas en duplicados: {len(duplicados):,}")
display(duplicados.sort_values('ID').head(10))

Total de filas involucradas en duplicados: 167,706


,ID,AÑO,SECTOR,GENERO,GRUPO ETAREO,REGION,DIAGNOSTICO,CAUSA EXTERNA
0,1,2024,PRIVADO,F,65+,INTERIOR,Hemorragia intracraneal,NC
1127284,1,2024,PRIVADO,F,65+,INTERIOR,Hemorragia intracraneal,NC
14,15,2024,PRIVADO,M,65+,INTERIOR,"Otros traumatismos de regiones especificadas, ...",Todas las demás causas externas
1116603,15,2024,PRIVADO,M,65+,INTERIOR,"Otros traumatismos de regiones especificadas, ...",Todas las demás causas externas
1094611,32,2024,PRIVADO,M,65+,INTERIOR,Otras enfermedades isquémicas del corazón,NC
31,32,2024,PRIVADO,M,65+,INTERIOR,Otras enfermedades isquémicas del corazón,NC
1114159,52,2024,PRIVADO,M,65+,INTERIOR,Fracturas de otros huesos de los miembros,Accidentes de transporte
51,52,2024,PRIVADO,M,65+,INTERIOR,Fracturas de otros huesos de los miembros,Accidentes de transporte
1068065,53,2024,PRIVADO,M,65+,INTERIOR,"Tumor maligno de la tráquea, de los bronquios ...",NC
52,53,2024,PRIVADO,M,65+,INTERIOR,"Tumor maligno de la tráquea, de los bronquios ...",NC


Elimino las duplicadas, manteniendo aquella que aparece primero y presento cuantificación de datos originales y posteriores.

In [18]:
df_limpio = df.drop_duplicates(keep='first')

print(f"Filas originales:  {len(df):,}")
print(f"Filas después:     {len(df_limpio):,}")
print(f"Filas eliminadas:  {len(df) - len(df_limpio):,}")

Filas originales:  1,132,011
Filas después:     1,048,158
Filas eliminadas:  83,853


En algunas columnas se presentas algunos datos que debo evaluar que hacer para limpiarlos: unificar o eliminar. Los presento y cuantifico.

In [19]:
print("- VALORES A LIMPIAR -")

print(f"\nREGION - nulos: {df_limpio['REGION'].isna().sum():,}")
print(f"REGION - 'Sin datos': {(df_limpio['REGION'] == 'Sin datos').sum():,}")
print(f"REGION - 'EXTERIOR': {(df_limpio['REGION'] == 'EXTERIOR').sum():,}")

print(f"\nGRUPO ETAREO - 'Sin datos': {(df_limpio['GRUPO ETAREO'] == 'Sin datos').sum():,}")

print(f"\nCAUSA EXTERNA - '99': {(df_limpio['CAUSA EXTERNA'] == '99').sum():,}")
print(f"CAUSA EXTERNA - 'SD': {(df_limpio['CAUSA EXTERNA'] == 'SD').sum():,}")
print(f"CAUSA EXTERNA - 'NC': {(df_limpio['CAUSA EXTERNA'] == 'NC').sum():,}")

- VALORES A LIMPIAR -

REGION - nulos: 39,597
REGION - 'Sin datos': 2,521
REGION - 'EXTERIOR': 269

GRUPO ETAREO - 'Sin datos': 2

CAUSA EXTERNA - '99': 11,192
CAUSA EXTERNA - 'SD': 1,769
CAUSA EXTERNA - 'NC': 947,609


De la anterior presentación y cuantificación identifico lo siguiente para cada columna:

REGION: unifico "nulos" y "sin datos" en "sin datos", que para el caso son lo mismo.

GRUPO ETAREO: la cantidad de "sin datos" es totalmente insignificante por lo que no hago cambios.

CAUSA EXTERNA: aquellos datos presentados como "SD" y "99" los unifico en "Sin datos" ya que a estos efectos significan lo mismo. Y los presentados como "NC", que significan "no corresponde", es decir que se entiende que no hubo una causa externa, son totalmente válidos por lo que los mantengo tal como estan.

Luego presento y cuantifico los datos de estas columnas.

In [20]:
# Elimino registros de EXTERIOR
df_limpio = df_limpio[df_limpio['REGION'] != 'EXTERIOR']
print(f"Filas después de eliminar EXTERIOR: {len(df_limpio):,}")

# Unifico nulos y 'Sin datos' en 'Sin datos'
df_limpio['REGION'] = df_limpio['REGION'].fillna('Sin datos')

# Unifico '99' y 'SD' en 'Sin datos'
df_limpio['CAUSA EXTERNA'] = df_limpio['CAUSA EXTERNA'].replace({'99': 'Sin datos', 'SD': 'Sin datos'})

print("- VERIFICACIÓN POST LIMPIEZA -")
print("\nREGION:")
print(df_limpio['REGION'].value_counts())
print("\nCAUSA EXTERNA:")
print(df_limpio['CAUSA EXTERNA'].value_counts())

Filas después de eliminar EXTERIOR: 1,008,292
- VERIFICACIÓN POST LIMPIEZA -

REGION:
REGION
INTERIOR      620930
MONTEVIDEO    384841
Sin datos       2521
Name: count, dtype: int64[pyarrow]

CAUSA EXTERNA:
CAUSA EXTERNA
NC                                                                   911631
Caídas                                                                31653
Todas las demás causas externas                                       26201
Accidentes de transporte                                              16059
Sin datos                                                             11739
Lesiones autoinflingidas intencionalmente                              6799
Agresiones                                                             3718
Envenenamiento accidental por, y exposición a, sustancias nocivas       368
Exposición al humo, fuego y llamas                                      123
Ahogamiento y sumersión accidentales                                      1
Name: count, dtype:

Entendiendo que el dataset original fue depurado, presento un resumen del final, antes de grabarlo en la sección solicitada.

In [21]:
print("- RESUMEN FINAL DEL DATASET LIMPIO -")
print(f"\nFilas originales:  1,132,011")
print(f"Filas finales:     {len(df_limpio):,}")
print(f"Columnas:          {df_limpio.shape[1]}")
print(f"\nValores nulos restantes:")
print(df_limpio.isnull().sum())

- RESUMEN FINAL DEL DATASET LIMPIO -

Filas originales:  1,132,011
Filas finales:     1,008,292
Columnas:          8

Valores nulos restantes:
ID               0
AÑO              0
SECTOR           0
GENERO           0
GRUPO ETAREO     0
REGION           0
DIAGNOSTICO      0
CAUSA EXTERNA    0
dtype: int64


Grabo el CSV limpio en data/processed/:

In [22]:
ruta_salida = os.path.join(base_dir, 'data', 'processed', 'egresos_hospitalarios_limpio.csv')

df_limpio.to_csv(ruta_salida, index=False, sep=';')

print(f"Archivo grabado en: {ruta_salida}")
print(f"Filas grabadas: {len(df_limpio):,}")

Archivo grabado en: C:\Users\usuario\PycharmProjects\nicolas_gallo_diploma_proyecto_python\data\processed\egresos_hospitalarios_limpio.csv
Filas grabadas: 1,008,292


La Fase 1 está completa:

- Carga del dataset (1,132,011 filas)
- Eliminación de 83,853 filas duplicadas
- Eliminación de los datos del exterior en REGION
- Unificación de nulos y 'Sin datos' en REGION
- Unificación de '99' y 'SD' en CAUSA EXTERNA
- Grabación de el CSV limpio en data/processed/